### PR TITLE
Add threads to Server io_service

### PR DESCRIPTION
**Pull request by ahesselgrave**
DelayHandler is a new class that should only be used for testing. Right now the delay is fixed to a const but if we really want we can change it to read from config. Honestly not worth it since it's only purpose is for the multithread_test.sh script.

### CODE REVIEW

**Comment by ahesselgrave** [comment]
son of a

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,7 @@
 # Project-wide compiler settings.
 export CXX=g++
 export CXXFLAGS=-c -std=c++11 -Wall -Wextra -Werror
-export LDFLAGS=-lgcov --coverage -lboost_system
+export LDFLAGS=-lgcov --coverage -lboost_system -lboost_thread
```
Comment (jobu95) [inline Makefile:11]: boost is so great

File: scripts/multithread_test.sh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+else
+  echo "Test passed."
+fi
+
+killall webserver
+
+rm $CONFIG_FILE
+rm $OUTPUT_FILE
+
+exit $EXIT_STATUS
```
Comment (jobu95) [inline scripts/multithread_test.sh:91]: LGTM!

File: src/config.cc [hunk 3]
```diff
@@ -7,160 +7,161 @@ using helpers::debugf;
...
-  for (auto const& stmt : conf->statements_) {
-    if (stmt->tokens_[0] == "path" || stmt->tokens_[0] == "default") {
-      info.push_back(stmt.get());
+    for (auto const& stmt : conf->statements_) {
+        if (stmt->tokens_[0] == "path" || stmt->tokens_[0] == "default") {
+            info.push_back(stmt.get());
+        }
     }
-  }
 }
```
Comment (jobu95) [inline src/config.cc:167]: I would say LGTM but honestly this function is so convoluted at this point that I have no idea what's going on anymore. If it passes unit tests I'm happy

Comment (jobu95) [reply src/config.cc:167]: Not to fault you, of course. We may want to consider refactoring some of this in the future

File: src/server.cc [hunk 4]
```diff
@@ -17,20 +18,48 @@ Server::Server(boost::asio::io_service& io_service, const NginxConfig *conf)
...
-        std::make_shared<Session>(std::move(socket_), conf_)->start();
-      	this->do_accept();
-      });
+  acceptor_.async_accept(
+    socket_,
+    boost::bind(
+      &Server::start_session,
+      this,
+      boost::asio::placeholders::error
+    ));
```
Comment (jobu95) [inline src/server.cc:64]: Thanks for the refactor - much cleaner.

**Comment by jobu95** [comment]
Looks good! Merging.

**Pull request merged**